### PR TITLE
M2 process data - Add data models

### PR DESCRIPTION
**Pull request by ram1117**
Hi, In this PR, I have implemented the following.

- Add `User`, `Post`, `Comment` and `Like` models.
- Add associations between the models as per the ERD diagram
- Add additional query methods to the models as per the requirements.

1.  `User` model - Added `last_three_posts` that returns last three posts for a given user.
2. `Post` model - Added `update_posts_count` method that updates post count of a user. Added `last_five_comments` method that returns last five comments for a given post.
3. `Comment` model - Added `update_comments_count` method that updates the comments count for that particular post.
4. `Like` model - Added `update_likes_count` method that updates the likes count for that particular post.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :likes
+  has_many :comments
+
+  def update_posts_count
+    author.update(posts_counter: author.posts.size)
+  end
+
```
Comment (elisha2kyakpo1) [inline app/models/post.rb]: - [ ] [OPTIONAL] We can make use of some model callback functions such as __after_save__.

```ruby
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter) # using the increment function will increment the posts_counter without calling on size function.
  end
```

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :post
+  belongs_to :author, class_name: 'User'
+
+  def update_likes_count
+    post.update(likes_counter: post.likes.size)
+  end
+end
```
Comment (elisha2kyakpo1) [inline app/models/like.rb]: - [ ] [OPTIONAL] We can make use of some model callback functions such as __after_save__.

```ruby
  after_save :update_likes_counter

  private

  def update_likes_counter
    author.increment!(:likes_counter) # using the increment function will increment the likes_counter without calling on size function.
  end
```

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :post
+  belongs_to :author, class_name: 'User'
+
+  def update_comments_count
+    post.update(comments_counter: post.comments.size)
+  end
+end
```
Comment (elisha2kyakpo1) [inline app/models/comment.rb]: - [ ] [OPTIONAL] We can make use of some model callback functions such as __after_save__.

```ruby
  after_save :update_comments_counter

  private

  def update_comments_counter
    author.increment!(:comments_counter) # using the increment function will increment the comments_counter without calling on size function.
  end
```

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 @ram1117, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

👍 Model associations were implemented properly, taking into account the 👏👏 __author_id__
✔️ The PR title and description are good and professional.
✔️ No linter errors.
✔️ Descriptive readme file.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 1 above)
Comment (ram1117) [reply app/models/post.rb]: @elisha2kyakpo1 thank you for these suggestions, I was unaware of the callback functions.

**Pull request merged**